README.md
# FetchPool

A library to easily and asynchronously download a list of URLs in parallel, using a given
maximum number of concurrent connections. This lets you download a list of hundreds of files,
while only ever downloading a few at the same time.

[![Pub](https://img.shields.io/pub/v/fetch_pool.svg)](https://pub.dartlang.org/packages/fetch_pool)
[![Tests](https://github.com/thesmythgroup/fetch_pool/actions/workflows/dart.yml/badge.svg)](https://github.com/thesmythgroup/fetch_pool/actions/workflows/dart.yml)

## Usage

A simple usage example:

```dart
import 'package:fetch_pool/fetch_pool.dart';

main() {
  const urls = [
    "https://picsum.photos/id/0/5616/3744",
    "https://picsum.photos/id/1/5616/3744",
    "https://some.invalid.url/to/simulate/an/error", // intentional
    "https://picsum.photos/id/1001/5616/3744",
    "https://picsum.photos/id/1002/4312/2868",
    "https://picsum.photos/id/1003/1181/1772",
    "https://picsum.photos/id/1004/5616/3744",
    "https://picsum.photos/id/1005/5760/3840",
  ];

  final pool = FetchPool(maxConcurrent: 2, urls: urls, destinationDirectory: "./deep/path/to/images");
  final results = await pool.fetch(progressCallback: (progress) {
    print('Total progress: $progress');
  });

  results.forEach((url, result) {
    if (result.isSuccess) {
      print('SUCCESS: ${url} > ${result.localPath}');
    } else {
      print('FAILURE: ${url} > ${result.error}');
    }
  });
}
```

The above code will try to download the list of URLs to the given `destinationDirectory`. It will use a
maximum of two concurrent connections and report on the total progress using the given `progressCallback`.

By default, the downloaded files will be named using the `FetchPoolFileNamingStrategy.basename` strategy.
That means that a URL like `https://test.com/img.png?a=123&b=456` will result in a local filename of `img.png`.
Using `basenameWithQueryParams` would result in `img_a_123_b_456.png`.
Using `base64EncodedUrl` base64 encodes the whole URL and would result in a local filename of `aHR0cHM6Ly90ZXN0LmNvbS9pbWcucG5nP2E9MTIzJmI9NDU2`.

By default, the `FetchPoolFileOverwritingStrategy.overwrite` strategy is used. That means that any existing files of the same name in the `destinationDirectory` will be overwritten. Using the `FetchPoolFileOverwritingStrategy.skip` strategy will skip existing files and not even attempt to download them. The `FetchPoolResult` object for each requested URL has a `persistenceResult` property that indicates if/how the file was persisted (`saved`, `overwritten`, or `skipped`).

## Credits

FetchPool is a project by [TSG](https://thesmythgroup.com/), a full-service digital agency taking software from concept to launch.
Our powerhouse team of designers and engineers build iOS, Android, and web apps across many industries.


CHANGELOG.md
# FetchPool Changelog

## 1.0.1

- Add `fileOverwritingStrategy` option of either `overwrite` (default) or `skip`. This enables a basic way to deal with the case that the destination file already exists on disk. The `FetchPoolResult` class has a new corresponding `persistenceResult` property of either `saved`, `overwritten`, or `skipped`.

## 1.0.0

- Initial version.


test/fetch_pool_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:fetch_pool/fetch_pool.dart';
import 'package:http/http.dart';
import 'package:http/testing.dart';
import 'package:test/test.dart';
import 'package:path/path.dart' as p;

void main() {
  group('Fetch Pool', () {
    final imageGifBytes = base64Decode(
        'R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7');
    String destinationDir = './__fetch_pool_test_images';

    tearDown(() {
      // Clean up files
      final dir = Directory(destinationDir);
      if (dir.existsSync()) {
        dir.deleteSync(recursive: true);
      }
    });

    test('filenameFromUrl', () {
      final url = 'https://test.com/img.png';
      final urlWithQuery = 'https://test.com/img.png?a=123&b=456';
      final urlWithoutExtension = 'https://test.com/img';
      final urlWithoutExtensionWithQuery = 'https://test.com/img?a=123&b=456';

      // basename (default)
      expect(FetchPool.filenameFromUrl(url), 'img.png');
      expect(FetchPool.filenameFromUrl(urlWithQuery), 'img.png');
      expect(FetchPool.filenameFromUrl(urlWithoutExtension), 'img');
      expect(FetchPool.filenameFromUrl(urlWithoutExtensionWithQuery), 'img');

      // basenameWithQueryParams
      expect(
          FetchPool.filenameFromUrl(
              url, FetchPoolFileNamingStrategy.basenameWithQueryParams),
          'img.png');
      expect(
          FetchPool.filenameFromUrl(urlWithQuery,
              FetchPoolFileNamingStrategy.basenameWithQueryParams),
          'img_a_123_b_456.png');
      expect(
          FetchPool.filenameFromUrl(urlWithoutExtension,
              FetchPoolFileNamingStrategy.basenameWithQueryParams),
          'img');
      expect(
          FetchPool.filenameFromUrl(urlWithoutExtensionWithQuery,
              FetchPoolFileNamingStrategy.basenameWithQueryParams),
          'img_a_123_b_456');

      // base64EncodedUrl
      expect(
          FetchPool.filenameFromUrl(
              url, FetchPoolFileNamingStrategy.base64EncodedUrl),
          'aHR0cHM6Ly90ZXN0LmNvbS9pbWcucG5n');
      expect(
          FetchPool.filenameFromUrl(
              urlWithQuery, FetchPoolFileNamingStrategy.base64EncodedUrl),
          'aHR0cHM6Ly90ZXN0LmNvbS9pbWcucG5nP2E9MTIzJmI9NDU2');
      expect(
          FetchPool.filenameFromUrl(urlWithoutExtension,
              FetchPoolFileNamingStrategy.base64EncodedUrl),
          'aHR0cHM6Ly90ZXN0LmNvbS9pbWc=');
      expect(
          FetchPool.filenameFromUrl(urlWithoutExtensionWithQuery,
              FetchPoolFileNamingStrategy.base64EncodedUrl),
          'aHR0cHM6Ly90ZXN0LmNvbS9pbWc_YT0xMjMmYj00NTY=');
    });

    test('Single Image', () async {
      final expectedDestinationPath = p.join(destinationDir, 'image.png');
      final fetchPool = FetchPool(
          maxConcurrent: 2,
          urls: ['http://example.com/image.png'],
          destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      final results = await fetchPool.fetch();
      expect(results.length, 1);
      expect(results.keys.first, 'http://example.com/image.png');
      final firstResult = results.values.first;
      expect(firstResult.url, 'http://example.com/image.png');
      expect(firstResult.error, null);
      expect(firstResult.localPath, expectedDestinationPath);

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, 1);
      expect(entries.first.path, expectedDestinationPath);
    });

    test('Multiple Images', () async {
      final urls =
          List.generate(50, (index) => 'http://example.com/image$index.png');
      final fetchPool = FetchPool(
          maxConcurrent: 5, urls: urls, destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      final results = await fetchPool.fetch();
      expect(results.length, urls.length);
      for (var url in urls) {
        final expectedDestinationPath = p.join(destinationDir, p.basename(url));
        final result = results[url];

        expect(result != null, true);
        expect(result!.url, url);
        expect(result.localPath, expectedDestinationPath);
      }

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, urls.length);
    });

    test('Multiple Images With Some Errors', () async {
      final urls =
          List.generate(50, (index) => 'http://example.com/image$index.png');
      final fetchPool = FetchPool(
          maxConcurrent: 5, urls: urls, destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        final filename = p.basename(request.url.path);

        if (filename == 'image10.png' || filename == 'image30.png') {
          return Response('Not found', 404);
        } else if (filename == 'image40.png') {
          return Response('Internal Server Error', 500);
        } else {
          return Response.bytes(imageGifBytes, 200);
        }
      });

      final results = await fetchPool.fetch();
      expect(results.length, urls.length);
      for (var url in urls) {
        final filename = p.basename(url);
        final expectedDestinationPath = p.join(destinationDir, filename);
        final result = results[url];

        expect(result != null, true);
        expect(result!.url, url);

        if (filename == 'image10.png' || filename == 'image30.png') {
          expect(result.localPath, null);
          expect(result.error, 'Status 404');
          expect(result.isSuccess, false);
        } else if (filename == 'image40.png') {
          expect(result.localPath, null);
          expect(result.error, 'Status 500');
          expect(result.isSuccess, false);
        } else {
          expect(result.localPath, expectedDestinationPath);
          expect(result.error, null);
          expect(result.isSuccess, true);
        }
      }

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, urls.length - 3);
    });

    test('Local filenames with basename fileNamingStrategy (default)',
        () async {
      final urls = List.generate(
          10, (index) => 'http://example.com/image$index.png?v=123&p=456');
      final fetchPool = FetchPool(
          maxConcurrent: 5, urls: urls, destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      final results = await fetchPool.fetch();
      final List<String> destinationPaths = [];
      expect(results.length, urls.length);

      for (var urlString in urls) {
        final url = Uri.parse(urlString);
        final expectedDestinationPath =
            p.join(destinationDir, p.basename(url.path));
        final result = results[urlString];

        destinationPaths.add(expectedDestinationPath);

        expect(result != null, true);
        expect(result!.url, urlString);
        expect(result.localPath, expectedDestinationPath);
      }

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, urls.length);

      for (var element in entries) {
        expect(destinationPaths.contains(element.path), true);
      }
    });

    test('Local filenames with basenameWithQueryParams fileNamingStrategy',
        () async {
      final urls = List.generate(
          10, (index) => 'http://example.com/image$index.png?v=123&p=456');
      final fetchPool = FetchPool(
          maxConcurrent: 5,
          urls: urls,
          destinationDirectory: destinationDir,
          fileNamingStrategy:
              FetchPoolFileNamingStrategy.basenameWithQueryParams);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      final results = await fetchPool.fetch();
      final List<String> destinationPaths = [];
      expect(results.length, urls.length);

      for (var urlString in urls) {
        final url = Uri.parse(urlString);
        final basenameWithoutExt = p.basenameWithoutExtension(url.path);
        final extension = p.extension(url.path);
        final expectedDestinationPath = p.join(
            destinationDir, '${basenameWithoutExt}_v_123_p_456$extension');
        final result = results[urlString];

        destinationPaths.add(expectedDestinationPath);

        expect(result != null, true);
        expect(result!.url, urlString);
        expect(result.localPath, expectedDestinationPath);
      }

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, urls.length);

      for (var element in entries) {
        expect(destinationPaths.contains(element.path), true);
      }
    });

    test('Local filenames with base64EncodedUrl fileNamingStrategy', () async {
      final urls = List.generate(
          10, (index) => 'http://example.com/image$index.png?v=123&p=456');
      final fetchPool = FetchPool(
          maxConcurrent: 5,
          urls: urls,
          destinationDirectory: destinationDir,
          fileNamingStrategy: FetchPoolFileNamingStrategy.base64EncodedUrl);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      final results = await fetchPool.fetch();
      final List<String> destinationPaths = [];
      expect(results.length, urls.length);

      for (var urlString in urls) {
        final base64EncodedUrl = base64Url.encode(utf8.encode(urlString));
        final expectedDestinationPath =
            p.join(destinationDir, base64EncodedUrl);
        final result = results[urlString];

        destinationPaths.add(expectedDestinationPath);

        expect(result != null, true);
        expect(result!.url, urlString);
        expect(result.localPath, expectedDestinationPath);
      }

      List<FileSystemEntity> entries =
          Directory(destinationDir).listSync(recursive: false).toList();
      expect(entries.length, urls.length);

      for (var element in entries) {
        expect(destinationPaths.contains(element.path), true);
      }
    });

    test('"overwrite" fileOverwritingStrategy (default)', () async {
      final urls =
          List.generate(10, (index) => 'http://example.com/image$index.png');
      final fetchPool = FetchPool(
          maxConcurrent: 5, urls: urls, destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      // Create three files with names that are about to be downloaded
      final existingFileIndexes = [3, 4, 5].toList();
      await Future.forEach(existingFileIndexes, (int fileIndex) async {
        String destinationPath = p.join(destinationDir, 'image$fileIndex.png');
        final file = File(destinationPath);
        await file.create(recursive: true);
        return file.writeAsString('Can\'t touch this');
      });

      final results = await fetchPool.fetch();
      expect(results.length, urls.length);

      // Check the persistenceResult values
      for (var url in results.keys) {
        if (['image3.png', 'image4.png', 'image5.png']
            .contains(p.basename(url))) {
          expect(results[url]!.persistenceResult,
              FetchPoolFilePersistenceResult.overwritten);
        } else {
          expect(results[url]!.persistenceResult,
              FetchPoolFilePersistenceResult.saved);
        }
      }

      // Make sure the files were overwritten
      for (var fileIndex in existingFileIndexes) {
        String destinationPath = p.join(destinationDir, 'image$fileIndex.png');
        final file = File(destinationPath);
        expect(file.readAsBytesSync(), imageGifBytes);
      }
    });

    test('"skip" fileOverwritingStrategy', () async {
      final urls =
          List.generate(10, (index) => 'http://example.com/image$index.png');
      final fetchPool = FetchPool(
          maxConcurrent: 5,
          urls: urls,
          destinationDirectory: destinationDir,
          fileOverwritingStrategy: FetchPoolFileOverwritingStrategy.skip);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      // Create three files with names that are about to be downloaded
      final existingFileIndexes = [3, 4, 5].toList();
      await Future.forEach(existingFileIndexes, (int fileIndex) async {
        String destinationPath = p.join(destinationDir, 'image$fileIndex.png');
        final file = File(destinationPath);
        await file.create(recursive: true);
        return file.writeAsString('Can\'t touch this');
      });

      final results = await fetchPool.fetch();
      expect(results.length, urls.length);

      // Check the persistenceResult values
      for (var url in results.keys) {
        if (['image3.png', 'image4.png', 'image5.png']
            .contains(p.basename(url))) {
          expect(results[url]!.persistenceResult,
              FetchPoolFilePersistenceResult.skipped);
        } else {
          expect(results[url]!.persistenceResult,
              FetchPoolFilePersistenceResult.saved);
        }
      }

      // Make sure the files not overwritten
      for (var fileIndex in existingFileIndexes) {
        String destinationPath = p.join(destinationDir, 'image$fileIndex.png');
        final file = File(destinationPath);
        expect(file.readAsStringSync(), 'Can\'t touch this');
      }
    });

    test('Total Progress Callback', () async {
      final urls =
          List.generate(50, (index) => 'http://example.com/image$index.png');
      final fetchPool = FetchPool(
          maxConcurrent: 5, urls: urls, destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        final filename = p.basename(request.url.path);

        if (filename == 'image10.png' || filename == 'image30.png') {
          return Response('Not found', 404);
        } else if (filename == 'image40.png') {
          return Response('Internal Server Error', 500);
        } else {
          return Response.bytes(imageGifBytes, 200);
        }
      });

      var progressInvocationCount = 0;
      double lastProgress = -1;
      await fetchPool.fetch(progressCallback: (progress) {
        expect(progress >= 0, true);
        expect(progress <= 100, true);
        expect(progress > lastProgress, true);
        lastProgress = progress;
        progressInvocationCount += 1;
      });

      expect(lastProgress, 100);
      expect(progressInvocationCount >= urls.length, true);
    });

    test('Ensure maxConcurrent is greater than 0', () async {
      expect(
          () =>
              FetchPool(maxConcurrent: 0, urls: [], destinationDirectory: '.'),
          throwsArgumentError);
      expect(
          () =>
              FetchPool(maxConcurrent: -1, urls: [], destinationDirectory: '.'),
          throwsArgumentError);
    });

    test('Ensure destinationDirectory is not empty', () async {
      expect(
          () => FetchPool(maxConcurrent: 1, urls: [], destinationDirectory: ''),
          throwsArgumentError);
      expect(
          () => FetchPool(
              maxConcurrent: 1, urls: [], destinationDirectory: '       '),
          throwsArgumentError);
    });

    test('Ensure the fetch method cannot be called twice', () async {
      final fetchPool = FetchPool(
          maxConcurrent: 2,
          urls: ['http://example.com/image.png'],
          destinationDirectory: destinationDir);
      fetchPool.client = MockClient((request) async {
        return Response.bytes(imageGifBytes, 200);
      });

      await fetchPool.fetch();

      // Calling fetch a second time should fail
      expect(() async => await fetchPool.fetch(), throwsStateError);
    });
  });
}


lib/src/fetch_pool.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:pool/pool.dart';
import 'package:path/path.dart' as p;

/// Result of fetching a single URL
class FetchPoolResult {
  /// URL that was (attempted to be) fetched
  final String url;

  /// Local path the URL was downloaded to
  final String? localPath;

  /// Local persistence result
  /// Was the file simply saved, overwritten, or skipped?
  final FetchPoolFilePersistenceResult? persistenceResult;

  /// Error in case of failure
  final Object? error;

  /// Indicates whether fetching was successful
  bool get isSuccess {
    return error == null;
  }

  /// Create a new instance
  FetchPoolResult(
      {required this.url, this.localPath, this.persistenceResult, this.error});
}

/// Enum describing the different file naming strategies.
enum FetchPoolFileNamingStrategy {
  /// Given a URL of https://test.com/img.png?a=123&b=456,
  /// results in a local filename of "img.png".
  basename,

  /// Given a URL of https://test.com/img.png?a=123&b=456,
  /// results in a local filename of "img_a_123_b_456.png".
  basenameWithQueryParams,

  /// Given a URL of https://test.com/img.png?a=123&b=456,
  /// results in a local filename of "aHR0cHM6Ly90ZXN0LmNvbS9pbWcucG5nP2E9MTIzJmI9NDU2".
  base64EncodedUrl
}

/// Enum describing the different file persistence results.
enum FetchPoolFilePersistenceResult {
  /// If the file didn't already exist in the target directory and was saved
  saved,

  /// If the file already existed in the target directory and was overwritten
  overwritten,

  /// If the file already existed in the target directory and was skipped
  skipped
}

/// Enum describing the different file overwriting strategies.
enum FetchPoolFileOverwritingStrategy {
  /// If the file already exists in the target directory, overwrite it
  overwrite,

  /// If the file already exists in the target directory, don't download it
  skip
}

/// Class used to fetch a list of URLs in parallel, only using a set number
/// of operations in parallel.
class FetchPool {
  /// Max number of concurrent operations (downloads)
  final int maxConcurrent;

  /// Destination download directory
  final String destinationDirectory;

  /// List of URLs to download
  final List<String> urls;

  /// Results keyed by the URL string
  final Map<String, FetchPoolResult> resultsByUrl = {};

  /// Indicates how the local filename should be named. Defaults to [basename].
  final FetchPoolFileNamingStrategy fileNamingStrategy;

  /// Indicates how to handle files of the same name that already exist in the
  /// target directory. Defaults to [overwrite].
  final FetchPoolFileOverwritingStrategy fileOverwritingStrategy;

  /// Used to prevent `fetch` from being called twice
  var _hasFetchBeenRun = false;

  /// The HTTP client to use
  ///
  /// This should usually only need to be changed for unit testing.
  http.Client client = http.Client();

  /// Creates a new instance allowing [maxConcurrent] parallel downloads
  ///
  /// If [destinationDirectory] doesn't exist, it will be created.
  FetchPool(
      {required this.maxConcurrent,
      required this.urls,
      required this.destinationDirectory,
      this.fileNamingStrategy = FetchPoolFileNamingStrategy.basename,
      this.fileOverwritingStrategy =
          FetchPoolFileOverwritingStrategy.overwrite}) {
    if (maxConcurrent < 1) {
      throw ArgumentError.value(maxConcurrent, 'maxConcurrent',
          'The maxConcurrent value must be greater than 0.');
    }

    if (destinationDirectory.trim().isEmpty) {
      throw ArgumentError.value(destinationDirectory, 'destinationDirectory',
          'The destinationDirectory must not be empty.');
    }
  }

  /// Convert the given Url string to a local filename, applying the given naming strategy
  ///
  /// See [FetchPoolFileNamingStrategy] for details.
  static String filenameFromUrl(String urlString,
      [FetchPoolFileNamingStrategy fileNamingStrategy =
          FetchPoolFileNamingStrategy.basename]) {
    final url = Uri.parse(urlString);
    switch (fileNamingStrategy) {
      case FetchPoolFileNamingStrategy.basename:
        return p.basename(url.path);
      case FetchPoolFileNamingStrategy.basenameWithQueryParams:
        final basenameWithoutExt = p.basenameWithoutExtension(url.path);
        final extension = p.extension(url.path);
        final convertedQueryString =
            '${url.hasQuery ? '_' : ''}${url.query.replaceAll(RegExp('&|=|\\?'), '_')}';
        return '$basenameWithoutExt$convertedQueryString$extension';
      case FetchPoolFileNamingStrategy.base64EncodedUrl:
        final bytes = utf8.encode(urlString);
        return base64Url.encode(bytes);
    }
  }

  /// Starts fetching the list of URLs
  ///
  /// Returns a [Future] with the map of results, keyed by each URL.
  /// This method is only allowed to be called once on a [FetchPool] instance.
  /// If you need to repeat a fetch, you need to create a fresh instance.
  ///
  /// A [progressCallback] function can optionally be passed in. It will
  /// be called repeatedly to report on the overall estimated progress. The progress
  /// is not calculated by the overall combined download size of all URLs (since
  /// that would require a roundtrip to the server for each URL before even beginning
  /// the download). Instead, the progress mainly reports the percentage of completed
  /// downloads plus the download percentage of active downloads.
  Future<Map<String, FetchPoolResult>> fetch(
      {Function(double)? progressCallback}) async {
    if (_hasFetchBeenRun) {
      throw StateError(
          'It is illegal to run fetch more than once on the same instance.');
    }

    _hasFetchBeenRun = true;

    var pool = Pool(maxConcurrent);
    final uniqueUrls = urls.toSet().toList();
    final Map<String, double> activeJobs = {};
    var completedJobCount = 0;
    double lastTotalProgress = 0;

    void notifyProgressCallback() {
      if (progressCallback != null) {
        final combinedActivePercentage =
            activeJobs.values.fold<double>(0, (sum, percent) => sum + percent);
        final activeJobFraction = activeJobs.isNotEmpty
            ? (activeJobs.length *
                (combinedActivePercentage / (activeJobs.length * 100)))
            : 0;
        final completedJobFraction = completedJobCount + activeJobFraction;
        final totalProgress = (completedJobFraction / uniqueUrls.length) * 100;

        if (totalProgress != lastTotalProgress) {
          progressCallback(totalProgress);
          lastTotalProgress = totalProgress;
        }
      }
    }

    var poolStream =
        pool.forEach<String, FetchPoolResult>(uniqueUrls, (urlString) async {
      final url = Uri.parse(urlString);
      final String filename = filenameFromUrl(urlString, fileNamingStrategy);
      final String destinationPath = p.join(destinationDirectory, filename);
      final File destinationFile = File(destinationPath);
      final bool destinationFileExists = await destinationFile.exists();
      FetchPoolResult result;

      double calculateProgress(int downloadedByteCount, int? contentLength) {
        if (contentLength != null && contentLength > 0) {
          return (downloadedByteCount / contentLength) * 100;
        }

        return 0;
      }

      activeJobs[urlString] = 0;

      if (fileOverwritingStrategy == FetchPoolFileOverwritingStrategy.skip &&
          destinationFileExists) {
        result = FetchPoolResult(
            url: urlString,
            localPath: destinationPath,
            persistenceResult: FetchPoolFilePersistenceResult.skipped);
      } else {
        final request = http.Request('GET', url);
        final http.StreamedResponse response = await client.send(request);

        if (response.statusCode == HttpStatus.ok) {
          await destinationFile.create(recursive: true);

          IOSink destinationFileWriteStream = destinationFile.openWrite();
          int downloadedByteCount = 0;

          await response.stream.listen((List<int> chunk) {
            // Display percentage of completion
            final percentage =
                calculateProgress(downloadedByteCount, response.contentLength);
            activeJobs[urlString] = percentage;
            notifyProgressCallback();

            destinationFileWriteStream.add(chunk);
            downloadedByteCount += chunk.length;
          }).asFuture();

          // Display percentage of completion
          final percentage =
              calculateProgress(downloadedByteCount, response.contentLength);
          activeJobs[urlString] = percentage;
          notifyProgressCallback();

          // Close the stream
          await destinationFileWriteStream.close();

          result = FetchPoolResult(
              url: urlString,
              localPath: destinationPath,
              persistenceResult: destinationFileExists
                  ? FetchPoolFilePersistenceResult.overwritten
                  : FetchPoolFilePersistenceResult.saved);
        } else {
          result = FetchPoolResult(
              url: urlString, error: 'Status ${response.statusCode}');
        }
      }

      activeJobs.remove(urlString);
      completedJobCount += 1;

      notifyProgressCallback();

      return result;
    }, onError: (urlString, error, stackTrace) {
      resultsByUrl[urlString] = FetchPoolResult(url: urlString, error: error);

      activeJobs.remove(urlString);
      completedJobCount += 1;

      notifyProgressCallback();

      /// Do not return the error to the output stream
      return false;
    });

    await for (var result in poolStream) {
      resultsByUrl[result.url] = result;
    }

    return resultsByUrl;
  }
}


example/fetch_pool_example.dart
import 'package:fetch_pool/fetch_pool.dart';

void main() async {
  const urls = [
    'https://picsum.photos/id/0/5616/3744',
    'https://picsum.photos/id/1/5616/3744',
    'https://picsum.photos/id/10/2500/1667',
    'https://picsum.photos/id/100/2500/1656',
    'https://picsumwursttest.photos/id/1000xAAA/5626/3635', // intentional
    'https://picsum.photos/id/1001/5616/3744',
    'https://picsum.photos/id/1002/4312/2868',
    'https://picsum.photos/id/1003/1181/1772',
    'https://picsum.photos/id/1004/5616/3744',
    'https://picsum.photos/id/1005/5760/3840',
    'https://picsum.photos/id/1006/3000/2000',
    'https://picsum.photos/id/1008/5616/3744',
    'https://picsum.photos/id/1009/5000/7502',
    'https://picsum.photos/id/101/2621/1747',
    'https://picsum.photos/id/1010/5184/3456',
    'https://picsum.photos/id/1011/5472/3648',
    'https://picsum.photos/id/1012/3973/2639',
    'https://picsum.photos/id/1013/4256/2832',
    'https://picsum.photos/id/1014/6016/4000',
    'https://picsum.photos/id/1015/6000/4000',
    'https://picsum.photos/id/1016/3844/2563',
    'https://picsum.photos/id/1018/3914/2935',
    'https://picsum.photos/id/1019/5472/3648',
    'https://picsum.photos/id/102/4320/3240',
    'https://picsum.photos/id/1020/4288/2848',
    'https://picsum.photos/id/1021/2048/1206',
    'https://picsum.photos/id/1022/6000/3376',
    'https://picsum.photos/id/1023/3955/2094',
    'https://picsum.photos/id/1024/1920/1280',
    'https://picsum.photos/id/1025/4951/3301'
  ];

  final pool = FetchPool(
      maxConcurrent: 2,
      urls: urls,
      destinationDirectory: './deep/path/to/images');
  final results = await pool.fetch(progressCallback: (progress) {
    print('Total progress: $progress');
  });

  results.forEach((url, result) {
    if (result.isSuccess) {
      print('SUCCESS: $url > ${result.localPath}');
    } else {
      print('FAILURE: $url > ${result.error}');
    }
  });

  print('Done');
}
